source/geod24/LocalRest.d

```diff
@@ -179,6 +179,11 @@ private final class LocalScheduler : C.FiberScheduler
     /// Request IDs waiting for a response
     private Waiting[ulong] waiting;
 
+    public this () @safe
```

Comment:
This is not necessary, it is generated by default

Reply:
> start should only every be called once though. Calling it twice is a programmer's error. So the check should probably be there (and be an assert failure).


That's a good suggestion.
There can be two exceptions:
The first is the second test code in LocalRest.d.
```
/// In a real world usage, users will most likely need to use the registry
unittest
{
    import std.conv;
    import geod24.concurrency;
    import geod24.Registry;

    __gshared Registry registry;
    registry.initialize();

    static interface API
    {
        @safe:
        public @property ulong pubkey ();
        public Json getValue (ulong idx);
        public string recv (Json data);
        public string recv (ulong index, Json data);

        public string last ();
    }

    static class Node : API
    {
        @safe:
        public this (bool isByzantine) { this.isByzantine = isByzantine; }
        public override @property ulong pubkey ()
        { lastCall = `pubkey`; return this.isByzantine ? 0 : 42; }
        public override Json getValue (ulong idx)
        { lastCall = `getValue`; return Json.init; }
        public override string recv (Json data)
        { lastCall = `recv@1`; return null; }
        public override string recv (ulong index, Json data)
        { lastCall = `recv@2`; return null; }

        public override string last () { return this.lastCall; }

        private bool isByzantine;
        private string lastCall;
    }

    static RemoteAPI!API factory (string type, ulong hash)
    {
        const name = hash.to!string;
        auto transceiver = registry.locate(name);
        if (transceiver !is null)
            return new RemoteAPI!API(transceiver);

        switch (type)
        {
        case "normal":
            auto ret =  RemoteAPI!API.spawn!Node(false);
            registry.register(name, ret.tid());
            return ret;
        case "byzantine":
            auto ret =  RemoteAPI!API.spawn!Node(true);
            registry.register(name, ret.tid());
            return ret;
        default:
            assert(0, type);
        }
    }

    Mutex mutex = new Mutex;
    Condition condition = new Condition(mutex);

    auto node1 = factory("normal", 1);
    auto node2 = factory("byzantine", 2);

    auto scheduler = new ThreadScheduler();
    scheduler.spawn({
        thisScheduler.start({
            auto node12 = factory("this does not matter", 1);
            auto node22 = factory("neither does this", 2);

            assert(node12.pubkey() == 42);
            assert(node12.last() == "pubkey");
            assert(node22.pubkey() == 0);
            assert(node22.last() == "pubkey");

            node12.recv(42, Json.init);
            assert(node12.last() == "recv@2");
            node12.recv(Json.init);
            assert(node12.last() == "recv@1");
            assert(node22.last() == "pubkey");

            synchronized (mutex) {
                condition.notify;
            }
        });
    });

    synchronized (mutex) {
        condition.wait(5000.msecs);
    }

    node1.ctrl.shutdown();
    node2.ctrl.shutdown();
}
```
If I remove 'dispatching' from the dispatch function,
I have to change the code as follows.

```
/// In a real world usage, users will most likely need to use the registry
unittest
{
    import std.conv;
    import geod24.concurrency;
    import geod24.Registry;

    __gshared Registry registry;
    registry.initialize();

    static interface API
    {
        @safe:
        public @property ulong pubkey ();
        public Json getValue (ulong idx);
        public string recv (Json data);
        public string recv (ulong index, Json data);

        public string last ();
    }

    static class Node : API
    {
        @safe:
        public this (bool isByzantine) { this.isByzantine = isByzantine; }
        public override @property ulong pubkey ()
        { lastCall = `pubkey`; return this.isByzantine ? 0 : 42; }
        public override Json getValue (ulong idx)
        { lastCall = `getValue`; return Json.init; }
        public override string recv (Json data)
        { lastCall = `recv@1`; return null; }
        public override string recv (ulong index, Json data)
        { lastCall = `recv@2`; return null; }

        public override string last () { return this.lastCall; }

        private bool isByzantine;
        private string lastCall;
    }

    static RemoteAPI!API factory (string type, ulong hash)
    {
        const name = hash.to!string;
        auto transceiver = registry.locate(name);
        if (transceiver !is null)
            return new RemoteAPI!API(transceiver);

        switch (type)
        {
        case "normal":
            auto ret =  RemoteAPI!API.spawn!Node(false);
            registry.register(name, ret.tid());
            return ret;
        case "byzantine":
            auto ret =  RemoteAPI!API.spawn!Node(true);
            registry.register(name, ret.tid());
            return ret;
        default:
            assert(0, type);
        }
    }

    Mutex mutex = new Mutex;
    Condition condition = new Condition(mutex);

    auto node1 = factory("normal", 1);
    auto node2 = factory("byzantine", 2);

    auto scheduler = new ThreadScheduler();
    scheduler.spawn({
            auto node12 = factory("this does not matter", 1);
            auto node22 = factory("neither does this", 2);

            assert(node12.pubkey() == 42);
            assert(node12.last() == "pubkey");
            assert(node22.pubkey() == 0);
            assert(node22.last() == "pubkey");

            node12.recv(42, Json.init);
            assert(node12.last() == "recv@2");
            node12.recv(Json.init);
            assert(node12.last() == "recv@1");
            assert(node22.last() == "pubkey");

            synchronized (mutex) {
                condition.notify;
            }
    });

    synchronized (mutex) {
        condition.wait(5000.msecs);
    }

    node1.ctrl.shutdown();
    node2.ctrl.shutdown();
}
```

I tried to increase efficiency by allowing all commands to be put in one FiberScheduler.start in the thread.

It has also previously been implemented to call FiberScheduler.start from the channel. Even at this time, ‘dispatching’ was necessary.

Reply:
@Geod24 Please confirm that it is a good idea to get rid of it.

Reply:
The following code is not currently part of the unittest, and is added by this PR:
```D
scheduler.spawn({
    thisScheduler.start({
```

Why is it needed ? If `thisScheduler` is a `FiberScheduler`, then it's not the same `start` that is being called.

Reply:
I think it's better not to make a big difference.
I remove 'dispatching' from the dispatch function.